core/src/main/java/com/linecorp/armeria/client/endpoint/LazyList.java

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.Supplier;
+
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableList;
+
+final class LazyList<E> implements List<E> {
```

Comment:
How about extends Guava's `ForwardingList` instead of implementing `List`?
I think we can reduce the overall code size with the same result. 😉
For example:
```java
final class LazyList<E> extends ForwardingList<E> {
    @Override
    protected List<E> delegate() {
        final List<E> delegate = this.delegate;
        if (delegate != null) {
            return delegate;
        }

        final List<E> supplied = ImmutableList.copyOf(delegateSupplier.get());
        if (delegateUpdater.compareAndSet(this, null, supplied)) {
            return supplied;
        } else {
            return this.delegate;
        }
    }

    // Override unwanted methods
    @Override
    public void add(int index, E element) {
        throw new UnsupportedOperationException();
    }
    ...
}

Reply:
That's a good suggestion. 👍

Reply:
@ikhoon Fixed. 😉 
FYI, I didn't override the mutable methods because they throw exceptions in the implementation anyway.

Reply:
>  FYI, I didn't override the mutable methods because they throw exceptions in the implementation anyway.

Ah. you are right. 🤣